### PR TITLE
WorkerSettings: Add disableLiburing option (enable_liburing in Rust)

### PR DESCRIPTION
**Pull request by ibc**
### Details

- Node: `createWorker({ disableLiburing: true })` disables liburing usage despite it's supported by the worker and current host.
- Rust: `create_worker({ enable_liburing: false })` disables liburing usage despite it's supported by the worker and current host.
- Related (still to be fixed) issue which brings lot of context: https://github.com/versatica/mediasoup/issues/1435

### CODE REVIEW

File: rust/src/worker.rs [hunk 1]
```diff
@@ -404,6 +410,10 @@ impl Inner {
             ));
         }
 
+        if let Some(disable_liburing) = disable_liburing {
```
Comment (ibc) [inline rust/src/worker.rs]: I assume here we are checking if user given `disable_liburing` is true.

File: rust/src/worker.rs [hunk 2]
```diff
@@ -192,6 +192,8 @@ pub struct WorkerSettings {
     /// "WebRTC-Bwe-AlrLimitedBackoff/Enabled/".
     #[doc(hidden)]
     pub libwebrtc_field_trials: Option<String>,
+    /// Disable liburing (io_uring) despite it's supported in current host.
+    pub disable_liburing: Option<bool>,
```
Comment (nazar-pc) [inline rust/src/worker.rs]: What about this instead?:
```suggestion
    pub disable_liburing: bool,
```

Or better yet (disable: false is much less natural):
```suggestion
    pub enable_liburing: bool,
```

Comment (ibc) [reply rust/src/worker.rs]: No please, here we want to keep current behavior and parity with Node layer. Default behavior must be that liburing is enabled if supported by current host.

And why should `disable_liburing` be mandatory? I want it to be optional.

Comment (nazar-pc) [reply rust/src/worker.rs]: Because by making it optional:
* you have tri-state boolean: Some(true), Some(false), None
* you don't let user know what the default value is

By having it as `bool` API is much clearer and self-documenting. Setting option to `false` is just as easy as setting it to `None` in `impl Default for WorkerSettings`.

On worker level it will be the same, but on Rust API level I would prefer `enable_io_uring` defaulting to `true`, which user can set to `false`. This is a clearer API, doesn't surface what library is used for `io_uring` and it is more natural to read `enable: true|false`.

File: rust/src/worker.rs (hunk 1 above)
Comment (nazar-pc) [reply rust/src/worker.rs]: No, you're checking that `disable_liburing` is `Some()`, for checking what is inside you'd also have to check the boolean `disable_liburing` itself. Can be simplified to `disable_liburing == Some(true)`. One more reason to not have tri-state boolean though.

File: rust/src/worker.rs (hunk 2 above)
Comment (ibc) [reply rust/src/worker.rs]: > on Rust API level I would prefer enable_io_uring defaulting to true

It's not defaulting to anything, you are making it mandatory for the user to pass true or false.

Comment (nazar-pc) [reply rust/src/worker.rs]: You may be misunderstanding Rust API. It is not making anything mandatory, the field always has to be specified anyway and you're setting it to `None` in `impl Default for WorkerSettings`. You can't not specify fields like in JavaScript. So you just change `Option<bool>` to `bool` and set default value to `false` instead of `None`.

The struct is intentionally marked as `#[non_exhaustive]`, so user can't construct it directly anyway, the only way is to go through APIs provided by the library like `impl Default for WorkerSettings`. This is also the reason why this is actually **not** a breaking change for library API!

Comment (ibc) [reply rust/src/worker.rs]: ok done: https://github.com/versatica/mediasoup/pull/1442/commits/0529246e7668f97542bf850e386434cd5ca6012b

**Comment by ibc** [comment]
OMG Rust tests fail because of this:

```
mediasoup-worker::mediasoup_worker_run() | settings error: unknown long option given as argument
```

I **hate** that `getopt.h` ancient thing...

And BTW it fails due to this obviously wrong code that I'm fixing now, but it doesn't justify the problem:

```rs
if enable_liburing {
    spawn_args.push("--disable_liburing".to_string());
}
```

File: rust/src/worker.rs [hunk 3]
```diff
@@ -404,6 +413,10 @@ impl Inner {
             ));
         }
 
+        if enable_liburing {
```
Comment (nazar-pc) [inline rust/src/worker.rs]:
```suggestion
        if !enable_liburing {
```

Comment (ibc) [reply rust/src/worker.rs]: yes yes yes, but see https://github.com/versatica/mediasoup/pull/1442#issuecomment-2278380704 please

File: rust/src/worker.rs [hunk 4]
```diff
@@ -404,6 +413,10 @@ impl Inner {
             ));
         }
 
+        if !enable_liburing {
+            spawn_args.push("--disable_liburing".to_string());
```
Comment (nazar-pc) [inline rust/src/worker.rs]: I believe it should have been the same as in settings file:
```suggestion
            spawn_args.push("--disableLiburing".to_string());
```

Comment (ibc) [reply rust/src/worker.rs]: OMG....

**Comment by ibc** [comment]
Ok, this is done.

**Review by nazar-pc — APPROVED** [review]
Assuming CI passes this looks fine to me

**Comment by ibc** [comment]
Why is this happening in CI????

https://github.com/versatica/mediasoup/actions/runs/10323061408/job/28579657794?pr=1442

`cargo clippy` works in my machine and we force Rust version!!!!

**Comment by ibc** [comment]
OMG I don't want to spend all the weekend with this.

**Comment by ibc** [comment]
Ok, `cargo clippy --all-targets -- -D warnings` fails in local too.

**Comment by nazar-pc** [comment]
I suspect you might be forcing a different version locally. It wants you to write this idiomatic Rust instead:
```rust
let settings = WorkerSettings {
    enable_liburing: false,
    ..WorkerSettings::default()
};
```

**Comment by ibc** [comment]
Why does it complain about this?

```
error: field assignment outside of initializer for an instance created with Default::default()
  --> rust/src/router/tests.rs:21:13
   |
21 |             settings.enable_liburing = false;
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: consider initializing the variable with `worker::WorkerSettings { enable_liburing: false, ..Default::default() }` and removing relevant reassignments
  --> rust/src/router/tests.rs:20:13
   |
20 |             let mut settings = WorkerSettings::default();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

This is the code:

```rust
worker_manager
        .create_worker({
            let mut settings = WorkerSettings::default();
            settings.enable_liburing = false;

            settings
        })
        .await
        .expect("Failed to create worker")
```

We do THE SAME in many other files such as here:

```rust
let worker = worker_manager
            .create_worker({
                let mut settings = WorkerSettings::default();

                settings.log_level = WorkerLogLevel::Debug;
                settings.log_tags = vec![WorkerLogTag::Info];
                settings.rtc_port_range = 0..=9999;
                settings.dtls_files = Some(WorkerDtlsFiles {
                    certificate: "tests/integration/data/dtls-cert.pem".into(),
                    private_key: "tests/integration/data/dtls-key.pem".into(),
                });
                settings.libwebrtc_field_trials =
                    Some("WebRTC-Bwe-AlrLimitedBackoff/Disabled/".to_string());
                settings.app_data = AppData::new(CustomAppData { bar: 456 });

                settings
            })
            .await
            .expect("Failed to create worker with custom settings");
```

**Comment by nazar-pc** [comment]
I suspect it has a heuristic about number of fields or just didn't learn to handle more complex cases, not sure which one of those two

**Comment by ibc** [comment]
Pushed, should be green now.

**Comment by nazar-pc** [comment]
Generally what it suggests does look a bit nicer and even shorter, but not always

**Comment by ibc** [comment]
OMG now tests failing...

**Comment by ibc** [comment]
WTH???

https://github.com/versatica/mediasoup/actions/runs/10323816619/job/28581996102?pr=1442
```
  process didn't exit successfully: `/home/runner/work/mediasoup/mediasoup/target/debug/deps/mediasoup-8f5b5278ff2f31a1` (signal: 11, SIGSEGV: invalid memory reference)
```

**Comment by nazar-pc** [comment]
Memory issues... again :confused:

**Comment by ibc** [comment]
> Memory corruption... again 😕

When did this happen in a test recently? I don't remember.

**Comment by ibc** [comment]
It fails consistently, this is depressing, no changes have been done that should affect this AT ALL:

- https://github.com/versatica/mediasoup/actions/runs/10323816619/job/28583849467?pr=1442
- https://github.com/versatica/mediasoup/actions/runs/10323816619/job/28600342140?pr=1442

**Comment by ibc** [comment]
No idea why the issue happens consistently but indeed it's the already know issue in Rust due to the separate usrsctp thread. I worked on this some months ago and there is an unfinished and very complex PR for which honestly I don't have time to revive now. I think I will disable these failing tests because we do know that the underlying code can fail. Well, all this assuming that the failing test is using two workers with pipe transports. If not, ignore all I said here.

**Comment by ibc** [comment]
Those are the ways Rust tests systematically fail. Since they run in parallel and taking into account that memory issues could make tests fail at any time once 

```
running 27 tests
test data_structures::tests::dtls_fingerprint ... ok
test ortc::tests::generate_router_rtp_capabilities_succeeds ... ok
test ortc::tests::generate_router_rtp_capabilities_too_many_codecs ... ok
test ortc::tests::generate_router_rtp_capabilities_unsupported ... ok
test ortc::tests::get_producer_rtp_parameters_mapping_get_consumable_rtp_parameters_get_consumer_rtp_parameters_get_pipe_consumer_rtp_parameters_succeeds ... ok
test ortc::tests::get_producer_rtp_parameters_mapping_unsupported ... ok
test router::audio_level_observer::tests::router_close_event ... ok
test router::active_speaker_observer::tests::router_close_event ... ok
test router::consumer::tests::transport_close_event ... ok
test router::consumer::tests::producer_close_event ... ok
test router::direct_transport::tests::router_close_event ... ok
test router::data_producer::tests::transport_close_event ... ok
test router::data_consumer::tests::transport_close_event ... ok
test router::data_consumer::tests::data_producer_close_event ... ok
test router::pipe_transport::tests::data_producer_close_is_transmitted_to_pipe_data_consumer ... ok
test router::plain_transport::tests::router_close_event ... ok
test router::producer::tests::transport_close_event ... ok
error: test failed, to rerun pass `-p mediasoup --lib`

Caused by:
  process didn't exit successfully: `/home/runner/work/mediasoup/mediasoup/target/debug/deps/mediasoup-8f5b5278ff2f31a1` (signal: 11, SIGSEGV: invalid memory reference)
Error: Process completed with exit code 101.
##[debug]Finishing: cargo test
```

```
running 27 tests
test data_structures::tests::dtls_fingerprint ... ok
test ortc::tests::generate_router_rtp_capabilities_succeeds ... ok
test ortc::tests::generate_router_rtp_capabilities_too_many_codecs ... ok
test ortc::tests::generate_router_rtp_capabilities_unsupported ... ok
test ortc::tests::get_producer_rtp_parameters_mapping_unsupported ... ok
test ortc::tests::get_producer_rtp_parameters_mapping_get_consumable_rtp_parameters_get_consumer_rtp_parameters_get_pipe_consumer_rtp_parameters_succeeds ... ok
test router::audio_level_observer::tests::router_close_event ... ok
test router::active_speaker_observer::tests::router_close_event ... ok
test router::consumer::tests::producer_close_event ... ok
test router::data_consumer::tests::data_producer_close_event ... ok
test router::data_consumer::tests::transport_close_event ... ok
test router::consumer::tests::transport_close_event ... ok
test router::data_producer::tests::transport_close_event ... ok
test router::direct_transport::tests::router_close_event ... ok
test router::plain_transport::tests::router_close_event ... ok
test router::pipe_transport::tests::data_producer_close_is_transmitted_to_pipe_data_consumer ... ok
test router::producer::tests::transport_close_event ... ok
error: test failed, to rerun pass `-p mediasoup --lib`

Caused by:
  process didn't exit successfully: `/home/runner/work/mediasoup/mediasoup/target/debug/deps/mediasoup-8f5b5278ff2f31a1` (signal: 11, SIGSEGV: invalid memory reference)
Error: Process completed with exit code 101.
```

```
running 27 tests
test data_structures::tests::dtls_fingerprint ... ok
test ortc::tests::generate_router_rtp_capabilities_succeeds ... ok
test ortc::tests::generate_router_rtp_capabilities_too_many_codecs ... ok
test ortc::tests::generate_router_rtp_capabilities_unsupported ... ok
test ortc::tests::get_producer_rtp_parameters_mapping_unsupported ... ok
test ortc::tests::get_producer_rtp_parameters_mapping_get_consumable_rtp_parameters_get_consumer_rtp_parameters_get_pipe_consumer_rtp_parameters_succeeds ... ok
test router::active_speaker_observer::tests::router_close_event ... ok
test router::audio_level_observer::tests::router_close_event ... ok
test router::consumer::tests::transport_close_event ... ok
test router::consumer::tests::producer_close_event ... ok
test router::data_consumer::tests::transport_close_event ... ok
test router::direct_transport::tests::router_close_event ... ok
test router::data_producer::tests::transport_close_event ... ok
test router::data_consumer::tests::data_producer_close_event ... ok
test router::plain_transport::tests::router_close_event ... ok
error: test failed, to rerun pass `-p mediasoup --lib`

Caused by:
  process didn't exit successfully: `/home/runner/work/mediasoup/mediasoup/target/debug/deps/mediasoup-8f5b5278ff2f31a1` (signal: 11, SIGSEGV: invalid memory reference)
Error: Process completed with exit code 101.
```

I would like to say that this is pipe transport related but the third test attempt doesn't even execute any "pipe" test.

**Comment by ibc** [comment]
This is completely amazing. Rust tests in v3 branch don't fail. But Rust tests in this PR branch fial consistently. Changes in this PR cannot be the culprit at all, they do not affect anything in Rust.

**Comment by ibc** [comment]
@nazar-pc help please? I doubt that Rust tests are even failing due to this well know issue:

https://github.com/versatica/mediasoup/issues/1352

In a comment above you can see that sometimes pipe transports do not even run and tests still fail. Maybe I'm wrong but AFAIK we only run more than 1 workers in Rust tests in those pipe transport related tests, so I'm completely lost.

**Comment by nazar-pc** [comment]
> Changes in this PR cannot be the culprit at all, they do not affect anything in Rust.

Either changes in this PR are the culprit (however unlikely that seems) or they trigger an issue that existed before as well, but was hard to reproduce.

> Maybe I'm wrong but AFAIK we only run more than 1 workers in Rust tests in those pipe transport related tests, so I'm completely lost.

There are as many tests running concurrently as CPU cores on the machine by default, so we have many workers with potentially different settings running at the same time and order is not deterministic.

File: worker/src/Settings.cpp [hunk 5]
```diff
@@ -73,13 +74,9 @@ void Settings::SetConfiguration(int argc, char* argv[])
 
 	optind = 1; // Set explicitly, otherwise subsequent runs will fail.
 	opterr = 0; // Don't allow getopt to print error messages.
+
 	while ((c = getopt_long_only(argc, argv, "", options, &optionIdx)) != -1)
 	{
-		if (!optarg)
-		{
-			MS_THROW_TYPE_ERROR("unknown configuration parameter: %s", optarg);
-		}
```
Comment (nazar-pc) [inline worker/src/Settings.cpp]: Why was this deleted? It seems like an important safety check that caller provided something meaningful as an input.

Comment (ibc) [reply worker/src/Settings.cpp]: Because the new `--disableLiburing` command line argument doesn't have any value so such a check throws if present. I can check that `optargs` exist for all the other arguments but didn't consider it necessary.

Comment (nazar-pc) [reply worker/src/Settings.cpp]: Well, but other options do require a value. Maybe we have a test that checks that and it causes memory corruption because you suddenly create a value out of null pointer?

**Comment by ibc** [comment]
> > Maybe I'm wrong but AFAIK we only run more than 1 workers in Rust tests in those pipe transport related tests, so I'm completely lost.
> 
> There are as many tests running concurrently as CPU cores on the machine by default, so we have many workers with potentially different settings running at the same time and order is not deterministic.

Do you mean that different tests run in parallel in the same Rust process? I mean, of course! Ok, I assume there is no way in Rust to create a separate process for each test file, right?

**Comment by nazar-pc** [comment]
> Do you mean that different tests run in parallel in the same Rust process? I mean, of course! Ok, I assume there is no way in Rust to create a separate process for each test file, right?

Yes, it does run in the same process by default. You could probably make it run in separate processes somehow, but we're dealing with memory safe language, so why, right? That is not a solution to the problem we have here anyway :slightly_smiling_face:

File: worker/src/Settings.cpp (hunk 5 above)
Comment (ibc) [reply worker/src/Settings.cpp]: I've made this change:

https://github.com/versatica/mediasoup/pull/1442/commits/db2252c4edc64a0f4be53f6d9d51622592ac25ba

Is it enough? What do you mean with "you suddenly create a value out of null pointer"? Command line arguments are created by Node and Rust layers in their `Worker` classes. Tests can not trigger wrong arguments passed to the worker.

Just wondering about this: In `worker/utils.rs`:

```rust
pub(super) fn run_worker_with_channels<OE>(
    id: WorkerId,
    thread_initializer: Option<Arc<dyn Fn() + Send + Sync>>,
    args: Vec<String>,
    worker_closed: Arc<AtomicBool>,
    on_exit: OE,
) -> WorkerRunResult
where
    OE: FnOnce(Result<(), ExitError>) + Send + 'static,
{
    let (channel, prepared_channel_read, prepared_channel_write) =
        Channel::new(Arc::clone(&worker_closed));
    let buffer_worker_messages_guard =
        channel.buffer_messages_for(SubscriptionTarget::String(std::process::id().to_string()));

    std::thread::Builder::new()
        .name(format!("mediasoup-worker-{id}"))
        .spawn(move || {
            if let Some(thread_initializer) = thread_initializer {
                thread_initializer();
            }
            let argc = args.len() as c_int;
            let args_cstring = args
                .into_iter()
                .map(|s| -> CString { CString::new(s).unwrap() })
                .collect::<Vec<_>>();
            let argv = args_cstring
                .iter()
                .map(|arg| arg.as_ptr().cast::<c_char>())
                .collect::<Vec<_>>();
            let version = CString::new(env!("CARGO_PKG_VERSION")).unwrap();

            let status_code = unsafe {
                let (channel_read_fn, channel_read_ctx, _channel_write_callback) =
                    prepared_channel_read.deconstruct();
                let (channel_write_fn, channel_write_ctx, _channel_read_callback) =
                    prepared_channel_write.deconstruct();

                mediasoup_sys::mediasoup_worker_run(
                    argc,
                    argv.as_ptr(),
                    version.as_ptr(),
                    0,
                    0,
                    channel_read_fn,
                    channel_read_ctx,
                    channel_write_fn,
                    channel_write_ctx,
                )
            };
```

Here `args` is a command line arguments string, something like:

```
"--logLevel=warn --disableLiburing"
```

Maybe something dangerous when doing this?:

```rust
let args_cstring = args
                .into_iter()
                .map(|s| -> CString { CString::new(s).unwrap() })
                .collect::<Vec<_>>();
            let argv = args_cstring
                .iter()
                .map(|arg| arg.as_ptr().cast::<c_char>())
                .collect::<Vec<_>>();
```

Comment (ibc) [reply worker/src/Settings.cpp]: Nope, no problem there IMHO. It just splits the string into these strings:
- "--logLevel=warn"
- "--disableLiburing"

It doesn't do anything like assuming/expecting a "=" symbol, so no danger here.

Comment (ibc) [reply worker/src/Settings.cpp]: I don't see why commit https://github.com/versatica/mediasoup/commit/db2252c4edc64a0f4be53f6d9d51622592ac25ba should fix this problem. It probably won't and, instead of wasting more time on this, I will change the new command line argument and add a value to it. No time to deal with ancient command line args stuff.

**Comment by ibc** [comment]
> You could probably make it run in separate processes somehow

No, trust me, I couldn't.

> but we're dealing with memory safe language, so why, right?

Because there is also C code running its own separate thread: https://github.com/versatica/mediasoup/issues/1352

**Comment by ibc** [comment]
Ok, some news:

- I can reproduce the Rust test failure in a real Linux host, no need to wait for CI.
- I've made command line argument mandatory here: https://github.com/versatica/mediasoup/pull/1442/commits/669ddca8a394bb985f191311689557990369ced3
- Now Rust tests fail even in macOS with this error:
  ```
  mediasoup-worker::mediasoup_worker_run() | settings error: missing value in command line 
  argument: (null)
  test router::data_producer::tests::transport_close_event ... ok
  test router::tests::worker_close_event ... FAILED
  ```

So it seems that we are sending some command line arg without value. I will print things to debug.

**Comment by ibc** [comment]
> Ok, some news:
> 
> * I can reproduce the Rust test failure in a real Linux host, no need to wait for CI.
> * I've made command line argument mandatory here: [669ddca](https://github.com/versatica/mediasoup/commit/669ddca8a394bb985f191311689557990369ced3)
> * Now Rust tests fail even in macOS with this error:
>   ```
>   mediasoup-worker::mediasoup_worker_run() | settings error: missing value in command line 
>   argument: (null)
>   test router::data_producer::tests::transport_close_event ... ok
>   test router::tests::worker_close_event ... FAILED
>   ```
> 
> So it seems that we are sending some command line arg without value. I will print things to debug.

Ignore, I forgot to add `=true` in Rust. Now things should work.

**Comment by ibc** [comment]
This is strongly depressing. Now WHAT????

```
running 27 tests
test data_structures::tests::dtls_fingerprint ... ok
test ortc::tests::generate_router_rtp_capabilities_succeeds ... ok
test ortc::tests::generate_router_rtp_capabilities_too_many_codecs ... ok
test ortc::tests::generate_router_rtp_capabilities_unsupported ... ok
test ortc::tests::get_producer_rtp_parameters_mapping_get_consumable_rtp_parameters_get_consumer_rtp_parameters_get_pipe_consumer_rtp_parameters_succeeds ... ok
test ortc::tests::get_producer_rtp_parameters_mapping_unsupported ... ok
test router::audio_level_observer::tests::router_close_event ... ok
test router::active_speaker_observer::tests::router_close_event ... ok
test router::data_consumer::tests::transport_close_event ... ok
test router::consumer::tests::producer_close_event ... ok
test router::consumer::tests::transport_close_event ... ok
test router::data_consumer::tests::data_producer_close_event ... ok
test router::data_producer::tests::transport_close_event ... ok
error: test failed, to rerun pass `-p mediasoup --lib`

Caused by:
  process didn't exit successfully: `/home/deploy/src/mediasoup-v3-deploy/target/debug/deps/mediasoup-8f5b5278ff2f31a1` (signal: 11, SIGSEGV: invalid memory reference)
```

**Comment by ibc** [comment]
I am 96% sure that the memory issue is here:

https://github.com/versatica/mediasoup/pull/1442/files#diff-5a9b1b4b01d5764dd21b1d1b7533d5795430dead60f84d450656db89da4e3930R18

**Comment by ibc** [comment]
Ohhh, so we are considering that liburing is enabled or disabled for all workers, but Rust tests run in **parallel** in same process and in just **one** of those tests we were **disabling** liburing. In the other ones liburing is enabled. So here the problem:

```c++
void DepLibUring::ClassInit()
{
	const auto mayor = io_uring_major_version();
	const auto minor = io_uring_minor_version();

	MS_DEBUG_TAG(info, "liburing version: \"%i.%i\"", mayor, minor);

	if (Settings::configuration.liburingDisabled)
	{
		MS_DEBUG_TAG(info, "liburing disabled by user settings");

		return;
	}

	// This must be called first.
	DepLibUring::CheckRuntimeSupport();

	if (DepLibUring::IsEnabled())
	{
		DepLibUring::liburing = new LibUring();

		MS_DEBUG_TAG(info, "liburing enabled");
	}
	else
	{
		MS_DEBUG_TAG(info, "liburing not enabled");
	}
}
```

- `DepLibUring::liburing` singleton is `thread_local static`.
- However we have `static bool enabled`.

File: rust/src/router/tests.rs [hunk 6]
```diff
@@ -16,7 +16,10 @@ async fn init() -> Worker {
     let worker_manager = WorkerManager::new();
 
     worker_manager
-        .create_worker(WorkerSettings::default())
+        .create_worker(WorkerSettings {
+            enable_liburing: false,
+            ..WorkerSettings::default()
+        })
```
Comment (ibc) [inline rust/src/router/tests.rs:22]: @nazar-pc, despite this works at intended (default settings are used but `enable_liburing` which is set to `false` instead), is this correct idiomatic syntax in Rust? 

In Node the default object must be added first, then those fields whose value must be different:

```ts
options = { ...options, foo: false }
```

It surprises me that in Rust it works if the changes are added first before the object with default values.

Comment (nazar-pc) [reply rust/src/router/tests.rs:22]: It is, I believe `..` should be used at the very end, I don't think it compiles otherwise at all

Comment (ibc) [reply rust/src/router/tests.rs:22]: Ok. For me idiomatic would be "here some values and after them those that override them" but it's ok. Rust guys have their own anti idiomatic concept of what idiomatic should mean.

File: worker/src/Settings.cpp (hunk 5 above)
Comment (nazar-pc) [reply worker/src/Settings.cpp]: I know we're not calling it incorrectly right now, but we could. And that would blow up instead of crashing with a nice message. Do not trust input, at least not to the degree that impacts memory safety.

Comment (ibc) [reply worker/src/Settings.cpp]: Now it's safe and we don't assume anything. Arg values are now mandatory. See latest changes.

**Comment by ibc** [comment]
Finally green. Merging!
